src/API.jl

```diff
@@ -261,175 +262,352 @@ function __installed(mode::PackageMode=PKGMODE_MANIFEST)
+        # Read the manifest in
         manifest = try
-            read_manifest(manifestfile)
+            read_manifest(path)
         catch e
-            @warn "Reading manifest file at $manifestfile failed with error" exception = e
-            nothing
+            @warn "Reading manifest file at $path failed with error" exception = e
+            return nothing
         end
-        manifest == nothing && continue
-        new_manifest_usage[manifestfile] = [Dict("time" => date)]
-        for (uuid, entry) in manifest
-            if entry.tree_hash !== nothing
-                push!(packages_to_keep,
-                      Operations.find_installed(entry.name, uuid, entry.tree_hash))
+
+        # Collect the locations of every package referred to in this manifest
+        pkg_dir(uuid, entry) = Operations.find_installed(entry.name, uuid, entry.tree_hash)
+        return [pkg_dir(u, e) for (u, e) in manifest if e.tree_hash != nothing]
+    end
+
+    function process_artifacts_toml(path)
+        # Not only do we need to check if this file doesn't exist, we also need to check
+        # to see if it this artifact is contained within a package that is going to go
+        # away.  This places an inherent ordering between marking packages and marking
+        # artifacts; the package marking must be done first so that we can ensure that
+        # all artifacts that are solely bound within such packages also get reaped.
+        if any(startswith(path, package_dir) for package_dir in packages_to_delete)
+            return nothing
+        end
+
+        artifact_dict = try
+            parse_toml(path)
+        catch e
+            @warn "Reading artifacts file at $path failed with error" exception = e
+            return nothing
+        end
+
+        artifact_path_list = String[]
+        for name in keys(artifact_dict)
+            getpaths(meta) = artifact_paths(SHA1(hex2bytes(meta["git-tree-sha1"])))
+            if isa(artifact_dict[name], Array)
+                for platform_meta in artifact_dict[name]
+                    append!(artifact_path_list, getpaths(platform_meta))
+                end
+            else
+                append!(artifact_path_list, getpaths(artifact_dict[name]))
+            end
+        end
+        return artifact_path_list
+    end
+
+    # Mark packages/artifacts as active or not by calling the appropriate
+    function mark(process_func::Function, index_files)
+        marked_paths = String[]
+        for index_file in index_files
+            # Check to see if it's still alive
+            paths = process_func(index_file)
+            if paths != nothing
+                # Print the path of this beautiful, extant file to the user
+                println("        $(Types.pathrepr(index_file))")
+                append!(marked_paths, paths)
+            end
+        end
+
+        # Return the list of absolutified, unique, marked paths
+        return Set(unique(marked_paths))
```

Comment:
Don't need `unique` if putting it in a `Set`

Reply:
Good catch; that was left over from some refactoring.